crates/red_knot_python_semantic/src/types/class.rs

```diff
@@ -1676,6 +1683,86 @@ impl<'db> From<ClassLiteralType<'db>> for Type<'db> {
     }
 }
 
+/// Representation of a single `Protocol` class definition.
+#[derive(Debug, Copy, Clone, PartialEq, Eq)]
+pub(super) struct ProtocolClassLiteral<'db>(ClassLiteralType<'db>);
+
+impl<'db> ProtocolClassLiteral<'db> {
+    pub(super) fn members(self, db: &'db dyn Db) -> &'db ordermap::set::Slice<Name> {
+        /// The list of excluded members is subject to change between Python versions,
+        /// especially for dunders, but it probably doesn't matter *too* much if this
+        /// list goes out of date. It's up-to-date as of Python commit 87b1ea016b1454b1e83b9113fa9435849b7743aa
+        /// (<https://github.com/python/cpython/blob/87b1ea016b1454b1e83b9113fa9435849b7743aa/Lib/typing.py#L1776-L1791>)
+        fn excluded_from_proto_members(member: &str) -> bool {
+            matches!(
+                member,
+                "_is_protocol"
+                    | "__non_callable_proto_members__"
+                    | "__static_attributes__"
+                    | "__orig_class__"
+                    | "__match_args__"
+                    | "__weakref__"
+                    | "__doc__"
+                    | "__parameters__"
+                    | "__module__"
+                    | "_MutableMapping__marker"
+                    | "__slots__"
+                    | "__dict__"
+                    | "__new__"
+                    | "__protocol_attrs__"
+                    | "__init__"
+                    | "__class_getitem__"
+                    | "__firstlineno__"
+                    | "__abstractmethods__"
+                    | "__orig_bases__"
+                    | "_is_runtime_protocol"
+                    | "__subclasshook__"
+                    | "__type_params__"
+                    | "__annotations__"
+                    | "__annotate__"
+                    | "__annotate_func__"
+                    | "__annotations_cache__"
+            )
+        }
+
+        #[salsa::tracked(return_ref)]
+        fn cached_members<'db>(
+            db: &'db dyn Db,
+            class: ClassLiteralType<'db>,
+        ) -> Box<ordermap::set::Slice<Name>> {
+            let mut members = FxOrderSet::default();
+
+            for parent_protocol in class
+                .iter_mro(db, None)
+                .filter_map(ClassBase::into_class)
+                .filter_map(|class| class.class_literal(db).0.into_protocol_class(db))
+            {
+                members.extend(
+                    symbol_table(db, parent_protocol.body_scope(db))
+                        .symbols()
+                        .filter(|symbol| symbol.is_bound() || symbol.is_declared())
```

Comment:
I don't think this will take visibility into account? I tried something similar for listing members of dataclasses at first, but then decided to iterate over declarations instead. You may want to do something similar here? See `ClassLiteralType::dataclass_fields`/`ClassLiteralType::own_dataclass_fields`

Reply:
The other advantage is that you'll get the type as well, using `symbol_from_declarations`.

Reply:
We should probably add a test for a protocol with a member defined under `if False` (or more realistically, `if sys.version_info >= ...` which evaluates to `False`).

Reply:
> I don't think this will take visibility into account? I tried something similar for listing members of dataclasses at first, but then decided to iterate over declarations instead.

Great point, thank you!! I did look at the methods you added for dataclasses, but only briefly, as I realised that I would need to include methods, properties and bindings-without-declarations in the list of protocol members, so I realised I wouldn't be able to simply reuse the methods you added for dataclasses. I should have looked more closely!